### ISSUE TITLE
Mac OSX Yosemite Install Problems and temp fix

### ISSUE THREAD

**Issue by luketaverne** [wnd-charm/wnd-charm#96]
Hello everyone,

First of all, thank you for producing such a great package, and the python wrappers for it. It's going to be very useful in an upcoming project.

I ran into some issues installing on OSX Yosemite. I've found a workaround that allowed me to build and get everything working properly, and I'll post it here for other users who may be experiencing this. When running `python setup.py build` I received the error
`gcc: warning: couldn’t understand kern.osversion ‘15.6.0` inside the `config.log` file. Running `which gcc` revealed that my gcc is in `/usr/local/bin`, and is actually gcc (not a sym link to clang). I investigated `/usr/bin/gcc`, but that is the Apple-given sym link to clang. After a few hours of trial and error, I came to the following solution:

1. Run `brew update` to make sure you have the latest packages (you need Homebrew installed for this)
2. `brew install gcc`. If brew tells you that it's already installed, run `brew upgrade gcc`. This should install gcc6.2.0 (at the time of posting). Then run `brew link gcc` to create the sym links in `/usr/local/bin`. You'll notice that these links are called `gcc6` instead of `gcc`. I chose to leave them like this, because I'm not building things with gcc that often. (I did append "OSX-OLD-10.9" to all of the defective `/usr/local/bin` gcc and c++ related things, so from now on gcc will refer to `/usr/bin/gcc`.)
3. Now we have to pass `gcc6`/`g++6` to the build script. The simplest way seems to be `CXX=g++-6 python setup.py build`. Everything should run smoothly this time!

If anyone has any input with regard to how this solution fits in to the big scheme of things, please let me know, I'd like to learn more about how these compiler issues work.

To the authors of the package: if other people start reporting this problem, maybe a suggestion could be added to the error parser/ console.log files to point them toward this solution? (i.e. if the first line of running `gcc` contains `gcc: warning: couldn’t understand kern.osversion ‘NN.N.N`, then we log an error saying if they are running OSX, maybe they should check that gcc is installed properly, and possibly point to this (or an improved and more robust) solution. If you'd like, I can try to make a pull request for this.

Thanks again everyone! I hope this helps someone.

**Comment by colettace** [comment]
Luke, thank you so much for this report. Seems like you had a homebrew installed version of GNU gcc at the beginning. Can you report back if you `brew unlink gcc` and try to compile if your build environment uses the Apple provided clang?

My development environment is Yosemite. When I run `gcc --version` the output is:
```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```

**Comment by luketaverne** [comment]
Yes, then it uses clang. It gives
```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 8.0.0 (clang-800.0.38)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
I tried this originally, and it was giving me some errors about C++ libraries not being linked (tr/unordered-map and so on). Now I built it again with gcc (apple clang) and it works fine. At one point, I managed to get it to compile with clang, and it produced this error when imported in Python:
```
Python 2.7.11 (default, Jun 19 2016, 19:21:39)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.72)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import wndcharm
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/wndcharm-0.9.14+dba43e8localmod-py2.7-macosx-10.11-x86_64.egg/wndcharm/__init__.py", line 27, in <module>
    from wndcharm import *
  File "/usr/local/lib/python2.7/site-packages/wndcharm-0.9.14+dba43e8localmod-py2.7-macosx-10.11-x86_64.egg/wndcharm/wndcharm.py", line 28, in <module>
    _wndcharm = swig_import_helper()
  File "/usr/local/lib/python2.7/site-packages/wndcharm-0.9.14+dba43e8localmod-py2.7-macosx-10.11-x86_64.egg/wndcharm/wndcharm.py", line 20, in swig_import_helper
    import _wndcharm
ImportError: dlopen(/usr/local/lib/python2.7/site-packages/wndcharm-0.9.14+dba43e8localmod-py2.7-macosx-10.11-x86_64.egg/_wndcharm.so, 2): Symbol not found: __ZNSs4_Rep11_S_terminalE
  Referenced from: /usr/local/lib/python2.7/site-packages/wndcharm-0.9.14+dba43e8localmod-py2.7-macosx-10.11-x86_64.egg/_wndcharm.so
  Expected in: flat namespace
 in /usr/local/lib/python2.7/site-packages/wndcharm-0.9.14+dba43e8localmod-py2.7-macosx-10.11-x86_64.egg/_wndcharm.so
```
Which makes me think my old homebrew installation was not completely removed the first time I tried with clang (old C++ libraries getting linked or something).

Unfortunately, it seems I'm unable to reproduce any of these problems now that the old gcc is sorted out completely...

**Comment by colettace** [comment]
You gotta blow away your old libraries. Try going into the WND-CHARM home directory and `rm -rf build/` then `python setup.py install` ... let me know if that fixes the python import problem....

If it doesn't, then what I think that means is that the binaries compiled by the two compilers are incompatible. I bet you're using a homebrew-installed Python, yes? If you used GNU gcc to compile it, it might not be able to interact with a clang compiled Python library in WND-CHARM. That happened to me when I migrated from using macports libraries compiled with GNU to homebrew libraries compiled with clang. In that case I literally recompiled every package. You might have to `brew uninstall --force python; brew cleanup` and then `brew install python` with gcc unlinked. Hopefully you don't have to recompile every package!

So I'm curious, why did you need GNU gcc on your mac? You have a specific homebrew package that needed gcc? AFAIK, compiling stuff with clang on Mas OS is best practice. Let me know how it goes!

**Comment by luketaverne** [comment]
> So I'm curious, why did you need GNU gcc on your mac?

Oh, probably some specific package I needed to build and install one time, and then never unlinked gcc after.

The second part to getting this problem fixed was manually removing the old and incorrectly built packages from the python directory (`/usr/local/lib/python2.7/site-packages/wnd-charm...` in my case). Luckily it worked without having to uninstall python, as my python version was mysteriously built with clang. I checked that gcc, g++ and c++ all gave the clang versions, then built and installed it this way and it works fine now. I also tried it with the g++-6 method I gave in the original post and was able to get it to work fine (although @colettace is right, just build it with clang. But this may be a useful diagnostic tool for someone: if g++-6 works fine, then you have old symlinks hidden somewhere that is mixing clang and gcc.)

Thanks again for the insight! I'm going to close this issue since it's not at all a problem with the package, but maybe some future user finds it useful in the archives. Thanks for a great tool!
